packages/solid/store/src/store.ts

```diff
@@ -235,141 +215,208 @@ export function updatePath(current: StoreNode, path: any[], traversed: (number |
   } else setProperty(current, part, value);
 }
 
-export type Readonly<T> = { readonly [K in keyof T]: DeepReadonly<T[K]> };
-export type DeepReadonly<T> = T extends [infer A]
-  ? Readonly<[A]>
-  : T extends [infer A, infer B]
-  ? Readonly<[A, B]>
-  : T extends [infer A, infer B, infer C]
-  ? Readonly<[A, B, C]>
-  : T extends [infer A, infer B, infer C, infer D]
-  ? Readonly<[A, B, C, D]>
-  : T extends [infer A, infer B, infer C, infer D, infer E]
-  ? Readonly<[A, B, C, D, E]>
-  : T extends [infer A, infer B, infer C, infer D, infer E, infer F]
-  ? Readonly<[A, B, C, D, E, F]>
-  : T extends [infer A, infer B, infer C, infer D, infer E, infer F, infer G]
-  ? Readonly<[A, B, C, D, E, F, G]>
-  : T extends [infer A, infer B, infer C, infer D, infer E, infer F, infer G, infer H]
-  ? Readonly<[A, B, C, D, E, F, G, H]>
-  : T extends object
-  ? Readonly<T>
-  : T;
+export type DeepReadonly<T extends Record<keyof any, any>> = {
+  readonly [K in keyof T]: T[K] extends NotWrappable ? T[K] : DeepReadonly<T[K]>;
+};
 
 export type StoreSetter<T> =
   | Partial<T>
-  | ((
-      prevState: T extends NotWrappable ? T : Store<DeepReadonly<T>>,
-      traversed?: (string | number)[]
-    ) => Partial<T | DeepReadonly<T>> | void);
+  | ((prevState: T, traversed?: (keyof any)[]) => Partial<T> | void);
+
 export type StorePathRange = { from?: number; to?: number; by?: number };
 
-export type ArrayFilterFn<T> = (
-  item: T extends any[] ? T[number] : never,
-  index: number
-) => boolean;
+export type ArrayFilterFn<T> = (item: T, index: number) => boolean;
 
-export type Part<T> = T extends any[]
-  ? keyof T | Array<keyof T> | ArrayFilterFn<T> | StorePathRange
-  : T extends object
-  ? keyof T | Array<keyof T>
-  : never;
+export type Part<T> = [T] extends [NotWrappable]
```

Comment:
Out of curiosity: do you have any article or could you tell me what should I search for to understand this, emm... thing?
`[T] extends [U]` 🤯

Reply:
Nvm. I think I found it https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#distributive-conditional-types :)

Reply:
I've removed this particular one. Since we're never going to pass a non-wrappable `T` to `Part`, there's no need to check and it improves performance which was previously a major concern.